### PR TITLE
Make game load units from JSON instead of CSV.

### PR DESCRIPTION
**Pull request by royfalk**
This affects main game and not save game (player ships).
This goes together with a change to asset-production that will include a unit.json file and conversion script.
This must be paired with https://github.com/vegastrike/Assets-Production/pull/59.

Code Changes:
- [x] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation

Issues:
- None identified.

### CODE REVIEW

File: engine/src/cmd/json.h [hunk 1]
```diff
@@ -0,0 +1,850 @@
...
+#include <cstdlib>
+#include <string>
+#include <vector>
+#include <cstdio>
+#include <utility>
+#include <stdexcept>
+#include <cctype>
+
+/*! \brief Base namespace for simpleson */
+namespace json
```
Comment (BenjamenMeyer) [inline engine/src/cmd/json.h:17]: Is there a reason we're implementing our own JSON parser?
Can't we just use Boost's JSON Parser (https://www.boost.org/doc/libs/1_79_0/boost/property_tree/json_parser.hpp)?

**Review by BenjamenMeyer — CHANGES_REQUESTED** [review]
Looks good; though I'd prefer to use an existing C/C++ JSON parser than implement our own.

File: engine/src/cmd/json.h (hunk 1 above)
Comment (stephengtuggy) [reply engine/src/cmd/json.h:17]: @royfalk are you using this library? https://github.com/gregjesl/simpleson

Comment (BenjamenMeyer) [reply engine/src/cmd/json.h:17]: That may make sense...but then we need to document it; however, that Boost::Json_Parser is more actively maintained and used by a larger audience, so less likely to have issues and we wouldn't have to maintain/sync it.

Comment (stephengtuggy) [reply engine/src/cmd/json.h:17]: I'm fine with using simpleson, personally. The code is already written to use that library. Why bother changing it? Also, I seem to recall reading that Boost's JSON parser leaves something to be desired.

We do need to figure out why the CI is failing, though.

File: engine/src/cmd/tests/json_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#include <gtest/gtest.h>
+
+#include "unit_json_factory.h"
+#include "unit_csv_factory.h"
+
+#include <iostream>
+#include <fstream>
+#include <iterator>
+
+#include <unistd.h>
```
Comment (stephengtuggy) [inline engine/src/cmd/tests/json_tests.cpp:35]: This header is not available on Windows systems. This is what is causing the Windows CI build to break, with the following error message:

```
D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\tests\json_tests.cpp(35,10): fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory [D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\build\Vega_Strike_tests.vcxproj]
```

**Comment by stephengtuggy** [comment]
On macOS, looks like the failure is that the JSON Sanity test fails with the following message:

```
C++ exception with description "Input was not an array" thrown in the test body.
```

**Comment by royfalk** [comment]
Interesting. The build fails because the unit test is unable to find the units.json file. However, I already merged the complementing assets PR. Does anyone have an idea why it's not picking up the file?

**Review by Loki1950 — APPROVED** [review]

**Comment by stephengtuggy** [comment]
Aw, man! The tests failed again. Perhaps the only thing missing is to copy the JSON file in CMakeLists.txt, as is done with another test file or two already.

File: engine/src/cmd/json.h (hunk 1 above)
Comment (stephengtuggy) [reply engine/src/cmd/json.h:17]: @BenjamenMeyer Issue filed as #686

**Review by stephengtuggy — APPROVED** [review]
LGTM!

**Review by Loki1950 — APPROVED** [review]

File: engine/src/cmd/json.h (hunk 1 above)
Comment (royfalk) [reply engine/src/cmd/json.h:17]: @BenjamenMeyer you will need to approve the PR, or I will not be able to merge.

Comment (BenjamenMeyer) [reply engine/src/cmd/json.h:17]: @stephengtuggy thanks! @royfalk will approve

**Comment by BenjamenMeyer** [comment]
@royfalk thanks - my last hold was waiting for the issue that @stephengtuggy filed; so all good now! Thanks!

**Pull request merged**